contributors/design-proposals/storage/container-storage-interface.md

```diff
@@ -0,0 +1,488 @@
+
+
+### CSI Credentials
+
+This part of proposal is not going to be implemented in alpha release.
+
+#### End user credentials
+CSI allows specifying *end user credentials* in all operations. Kubernetes does not have facility to configure a Secret per *user*, we usually track objects per *namespace*. Therefore we decided to postpone implementation of these credentials and wait until CSI is clarified.
+
+#### Volume specific credentials
+Some storage technologies (e.g. iSCSI with CHAP) require credentials tied to the volume (iSCSI LUN) that must be used during `NodePublish` request. It is expected that these credentials will be provided during dynamic provisioning of the volume, however CSI `CreateVolume` response does not provide any. In case it gets fixed soon external provisioner can save the secrets in a dedicated namespace and make them available to external attacher and internal CSI volume plugin using these `CSIPersistentVolumeSource` fields:
+
+// ...
+```go
+type CSIPersistentVolumeSource struct {
+
+    // Optional: MountSecretRef is a reference to the secret object containing
+    // sensitive information to pass to the CSI driver during NodePublish.
+    // This may be empty if no secret is required. If the secret object contains
+    // more than one secret, all secrets are passed.
+    // +optional
+    MountSecretRef *SecretReference `json:"mountSecretRef,omitempty" protobuf:"bytes,3,opt,name=mountSecretRef"`
+
+    // Optional: AttachSecretRef is a reference to the secret object containing
+    // sensitive information to pass to the CSI driver during ControllerPublish.
+    // This may be empty if no secret is required. If the secret object contains
+    // more than one secret, all secrets are passed.
+    // +optional
+    AttachSecretRef *SecretReference `json:"attachSecretRef,omitempty" protobuf:"bytes,4,opt,name=attachSecretRef"`
+}
+```
+
+Note that a malicious provisioner could obtain an arbitrary secret by setting the mount secret in PV object to whatever secret it wants. It is assumed that cluster admins will only run trusted provisioners.
+
+Because the kubelet would be responsible for fetching and passing the mount secret to the CSI driver,the Kubernetes NodeAuthorizer must be updated to allow kubelet read access to mount secrets.
+
+## Alternatives Considered
+
+### Extending PersistentVolume Object
+
+Instead of creating a new `VolumeAttachment` object, another option we considered was extending the exiting `PersistentVolume` object. 
+
+`PersistentVolumeSpec` would be extended to include:
+* List of nodes to attach the volume to (initially empty).
+
+`PersistentVolumeStatus` would be extended to include:
+* List of nodes the volume was successfully attached to.
+
+We dismissed this approach because having attach/detach triggered by the creation/deletion of an object is much easier to manage (for both external-attacher and Kubernetes) and more robust (fewer corner cases to worry about).
```

Comment:
What happens if I delete the PV before the attach objects, then a PV gets recreated with the same name?

Reply:
+@jsafrane there will be a finalizer on the PV to prevent deletion until the detach is completed.

Reply:
If PV is deleted before `AttachedVolume` is created, then external attacher controller won't even start attaching the volume (as there is nothing to attach).

When the PV is deleted *after* external attacher controller started attaching, the attacher already has finalizer on the PV - it puts the finalizer there before it starts attaching.